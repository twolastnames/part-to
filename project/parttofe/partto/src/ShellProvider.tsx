import React, { MutableRefObject, useRef, useState } from "react";

import { MantineProvider, createTheme } from "@mantine/core";
import { PropsWithChildren } from "react";
import "@mantine/core/styles.layer.css";
import "./App.scss";

const themes = {
  default: createTheme({
    primaryColor: "red",
  }),
  dark: createTheme({
    primaryColor: "red",
  }),
};

let themeChangeRef: MutableRefObject<(theme: Theme) => void>;

export enum Theme {
  Default = "default",
  Dark = "dark",
}

export function changeTheme(theme: Theme) {
  const themeNode = document.getElementById("theme");
  if (!themeNode || !themeNode.className) {
    console.error("developers should have created an theme ID in the project");
    return;
  }
  themeChangeRef?.current && themeChangeRef.current(theme);
  themeNode.className = `${theme}Theme`;
}

export const ShellProvider = ({ children }: PropsWithChildren) => {
  const [theme, setTheme] = useState<Theme>(Theme.Default);
  themeChangeRef = useRef(setTheme);
  return (
    <div id="theme" className="defaultTheme">
      <MantineProvider theme={themes[theme]}>{children}</MantineProvider>
    </div>
  );
};
