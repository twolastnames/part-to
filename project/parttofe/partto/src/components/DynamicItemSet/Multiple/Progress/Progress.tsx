import React from "react";

import classes from "./Progress.module.scss";
import { RingProgress } from "@mantine/core";

export interface ProgressProps {
  on: number;
  total: number;
  label: string;
  title: string;
  onClick: () => void;
}

export function Progress({ on, total, label, title, onClick }: ProgressProps) {
  return (
    <span data-testid="LocalProgress">
      <RingProgress
        size={56}
        thickness={8}
        onClick={onClick}
        title={title}
        label={
          <div onClick={onClick} className={classes.label}>
            {label}
          </div>
        }
        sections={[
          { value: ((on + 1) / total) * 100, color: "var(--secondary-color)" },
        ]}
        data-testid="Progress"
      />
    </span>
  );
}
