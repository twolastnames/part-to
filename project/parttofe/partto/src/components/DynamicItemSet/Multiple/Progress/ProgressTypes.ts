export interface ProgressProps {
  on: number;
  total: number;
  label: string;
  title: string;
  onClick: () => void;
}
