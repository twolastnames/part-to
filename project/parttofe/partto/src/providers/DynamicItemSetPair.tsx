import React, {
  createContext,
  useState,
  Context,
  MutableRefObject,
  createRef,
  useContext,
  SetStateAction,
} from "react";
import { PropsWithChildren } from "react";
import { Duration, getDuration } from "../shared/duration";
import { debounce, DebouncedFunc } from "lodash";

export type DynamicItemSetPairType = {
  selected: number;
  setSelected: (selected: number) => void;
  getTotal: () => number;
  goBack: () => void;
  goForward: () => void;
  showDuration: Duration;
  togglePause: () => void;
  paused: boolean;
  nextShowDuration: () => void;
};

export const UndefinedDynamicItemSetPair = {
  selected: 0,
  setSelected: (_: number) => undefined,
  goBack: () => undefined,
  goForward: () => undefined,
  getTotal: () => 0,
  showDuration: getDuration(0),
  togglePause: () => undefined,
  paused: true,
  nextShowDuration: () => undefined,
};

export function useSingleOfPair({ context }: ContextDescription) {
  return useContext(context) || UndefinedDynamicItemSetPair;
}

type ContextType = Context<undefined | DynamicItemSetPairType>;

export type ContextDescription = {
  context: ContextType;
  setCount: (value: number) => void;
  setSettingKey: (settingKey: string) => void;
  setDurations: (value: Array<Duration>) => void;
};

const defaultDurations = [1000, 2000, 4000, 8000];

const getDurations = (key: string) =>
  JSON.parse(
    localStorage.getItem(`${key}Durations`) || JSON.stringify(defaultDurations),
  ).map(getDuration);

const setDurations = (key: string, value: Array<Duration>) => {
  localStorage.setItem(
    `${key}Durations`,
    JSON.stringify(value.map((duration) => duration.toMilliseconds())),
  );
};

const leftCount: MutableRefObject<((value: number) => void) | null> =
  createRef();
const leftSettingKey: MutableRefObject<string | null> = createRef();
const getLeftSettingKey = () => leftSettingKey.current || "left";
function setLeftCount(value: number) {
  leftCount.current && leftCount.current(value);
}
const LeftDynamicItemSetPairContext: ContextType = createContext<
  undefined | DynamicItemSetPairType
>(undefined);

export const LeftContext: ContextDescription = {
  context: LeftDynamicItemSetPairContext,
  setCount: setLeftCount,
  setSettingKey: (value: string) => {
    leftSettingKey.current = value;
  },
  setDurations: (value: Array<Duration>) => {
    setDurations(getLeftSettingKey(), value);
  },
};

const rightCount: MutableRefObject<((value: number) => void) | null> =
  createRef();
const rightSettingKey: MutableRefObject<string | null> = createRef();
const getRightSettingKey = () => leftSettingKey.current || "left";

function setRightCount(value: number) {
  rightCount.current && rightCount.current(value);
}
const RightDynamicItemSetPairContext: ContextType = createContext<
  undefined | DynamicItemSetPairType
>(undefined);

export const RightContext: ContextDescription = {
  context: RightDynamicItemSetPairContext,
  setCount: setRightCount,
  setSettingKey: (value: string) => {
    rightSettingKey.current = value;
  },
  setDurations: (value: Array<Duration>) => {
    setDurations(getRightSettingKey(), value);
  },
};

type GoADirection = (
  totalPages: number,
  setSelected: (mutator: (arg: ProviderState) => ProviderState) => void,
) => void;

const goBack: GoADirection = (length, setState) => {
  setState((last: ProviderState) => ({
    ...last,
    selected: last.selected <= 0 ? length - 1 : last.selected - 1,
  }));
};

const goForward: GoADirection = (length, setState) => {
  setState((last: ProviderState) => ({
    ...last,
    selected: last.selected >= length - 1 ? 0 : last.selected + 1,
  }));
};

type ProviderState = {
  selected: number;
  showDuration: number;
  paused: boolean;
  count: number;
};

function useProvider({
  setCount,
  getSettingKey,
  togglePauseClickSlower,
}: {
  getSettingKey: () => string;
  setCount: MutableRefObject<((value: number) => void) | null>;
  togglePauseClickSlower: DebouncedFunc<
    (setState: (arg: SetStateAction<ProviderState>) => void) => void
  >;
}) {
  const durations = getDurations(getSettingKey());
  const [{ count, selected, showDuration, paused }, setState] =
    useState<ProviderState>({
      selected: 0,
      showDuration: Math.floor(durations.length / 2),
      paused: false,
      count: 0,
    });

  setCount.current = (value: number) => {
    if (value === count) {
      return;
    }
    setState((previous) => ({ ...previous, count: value }));
  };

  if (selected !== 0 && (selected < 0 || selected >= count)) {
    setState((current) => ({ ...current, selected: 0 }));
  }

  const duration = durations[showDuration];
  return {
    selected: selected,
    setSelected: (selected: number) => {
      setState((current) => ({ ...current, selected }));
    },
    goBack: () => {
      goBack(count, setState);
    },
    goForward: () => {
      goForward(count, setState);
    },
    showDuration: duration,
    paused: paused,
    getTotal: () => count,
    togglePause: () => {
      togglePauseClickSlower(setState);
    },
    nextShowDuration: () => {
      const next = showDuration + 1;
      setState((previous) => ({
        ...previous,
        showDuration: next >= durations.length ? 0 : next,
      }));
    },
  };
}

const leftTogglePauseClickSlower = debounce(
  (setState: (arg: SetStateAction<ProviderState>) => void) => {
    setState((previous) => ({ ...previous, paused: !previous.paused }));
  },
  100,
);

export function LeftDynamicItemSetPairProvider({
  children,
}: PropsWithChildren) {
  return (
    <LeftDynamicItemSetPairContext.Provider
      value={useProvider({
        setCount: leftCount,
        getSettingKey: getLeftSettingKey,
        togglePauseClickSlower: leftTogglePauseClickSlower,
      })}
    >
      {children}
    </LeftDynamicItemSetPairContext.Provider>
  );
}

const rightTogglePauseClickSlower = debounce(
  (setState: (arg: SetStateAction<ProviderState>) => void) => {
    setState((previous) => ({ ...previous, paused: !previous.paused }));
  },
  100,
);

export function RightDynamicItemSetPairProvider({
  children,
}: PropsWithChildren) {
  return (
    <RightDynamicItemSetPairContext.Provider
      value={useProvider({
        setCount: rightCount,
        getSettingKey: getRightSettingKey,
        togglePauseClickSlower: rightTogglePauseClickSlower,
      })}
    >
      {children}
    </RightDynamicItemSetPairContext.Provider>
  );
}
