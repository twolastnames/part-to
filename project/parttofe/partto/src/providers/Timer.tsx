import React, {
  createContext,
  ReactNode,
  useContext,
  useEffect,
  useMemo,
  useRef,
  useState,
} from "react";
import { PropsWithChildren } from "react";
import { DateTime, getDateTime } from "../shared/dateTime";
import { Duration, getDuration } from "../shared/duration";
import { RunState, TaskDefinitionId } from "../api/sharedschemas";
import { Timer } from "../components/Timer/Timer";
import {
  addAlarmNote,
  removeNote,
} from "../components/Layout/NavigationBar/Noted/Noted";
import {
  addRunStateListener,
  removeRunStateListener,
} from "../shared/runStateMessage";

type TimerDescription = {
  component: ReactNode;
  duration: Duration;
  started: DateTime;
  message: string;
  enforced: boolean;
  task: TaskDefinitionId;
  setOnLocate: (callback: () => void) => void;
};

export type TimerType = {
  timers: { [id: string]: TimerDescription };
};

export const TimerContext = createContext<undefined | TimerType>(undefined);

type TaskKeyed = { task: TaskDefinitionId };

type TimerFactory<T> = (args: T) => {
  component: ReactNode;
  duration: Duration;
  task: TaskDefinitionId;
};

function getMappedTimers<T extends TaskKeyed>(
  getTimer: TimerFactory<T>,
  timers: Array<T> | undefined,
): { [id: string]: TimerDescription } {
  return (
    timers?.reduce(
      (current, timer) => ({
        ...current,
        [timer.task]: getTimer(timer),
      }),
      {},
    ) || ({} as { [id: TaskDefinitionId]: T })
  );
}

export function TimerProvider({ children }: PropsWithChildren) {
  const [runState, setRunState] = useState<RunState | undefined>();
  const offsets = useRef<{
    [id: string]: Duration;
  }>({});
  const onLocates = useRef<{ [task: string]: () => void }>({});
  const audio = useRef(new Audio(require("./messageAlert.mp3")));
  const timestamp = runState?.timestamp;
  const timers = useMemo(
    () => ({
      ...getMappedTimers(
        ({ task, started, duration }) => ({
          task,
          started,
          setOnLocate: (callback: () => void) => {
            onLocates.current[task] = callback;
          },
          duration,
          enforced: true,
          message: "Duty Complete?",
          component: (
            <Timer
              start={started}
              duration={duration}
              adjustment={{
                offset: offsets.current[task] || getDuration(0),
                setOffset: (value: Duration) => (offsets.current[task] = value),
              }}
            />
          ),
        }),
        runState?.timers.enforced || [],
      ),
      ...getMappedTimers(
        ({ task, started, duration }) => ({
          task,
          started,
          setOnLocate: (callback: () => void) => {
            onLocates.current[task] = callback;
          },
          duration,
          enforced: false,
          message: "Past Expected Time",
          component: <Timer start={started} duration={duration} />,
        }),
        runState?.timers.laxed || [],
      ),
      ...getMappedTimers(
        ({ task, till }) => ({
          task,
          started: timestamp,
          setOnLocate: (callback: () => void) => {
            onLocates.current[task] = callback;
          },
          duration: till,
          enforced: true,
          message: "Need to Start Duty",
          component: <Timer start={timestamp} duration={till} />,
        }),
        runState?.timers.imminent || [],
      ),
    }),
    [runState, offsets, timestamp],
  );

  useEffect(() => {
    addRunStateListener(setRunState);
    return () => {
      removeRunStateListener(setRunState);
    };
  }, [setRunState]);

  useEffect(() => {
    const id = setInterval(() => {
      const now = getDateTime();
      for (const [key, timer] of Object.entries(timers)) {
        const expireAt = timer.started
          .add(offsets.current[key] || getDuration(0))
          .add(timer.duration);
        if (!timer.enforced || now.sinceEpoch() < expireAt.sinceEpoch()) {
          continue;
        }
        addAlarmNote({
          key,
          heading: "Alarm Expired",
          detail: timer.message,
          onClick: () => {
            onLocates.current[key]?.();
          },
        });
        if (!audio.current.paused) {
          return;
        }
        try {
          audio.current.play();
        } catch (NotAllowedError) {
          console.error(
            "could not start an alarm... likely because the page has not been interacted with",
          );
        }
        audio.current.loop = true;
        return;
      }
      for (const key of Object.keys(timers)) {
        removeNote(key);
      }
      audio.current.pause();
    }, 1000);
    return () => {
      clearInterval(id);
    };
  }, [offsets, runState, timers]);

  return (
    <TimerContext.Provider
      value={{
        timers,
      }}
    >
      {children}
    </TimerContext.Provider>
  );
}

export function useTimerProvider({
  task,
  onLocate,
}: {
  task: TaskDefinitionId;
  onLocate?: () => void;
}) {
  const timers = useContext(TimerContext)?.timers;
  onLocate && timers?.[task]?.setOnLocate(onLocate);
  return timers?.[task]?.component || <></>;
}
