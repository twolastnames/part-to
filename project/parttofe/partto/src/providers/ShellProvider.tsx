import React, { MutableRefObject, useRef, useState } from "react";

import { MantineProvider, createTheme } from "@mantine/core";
import { PropsWithChildren } from "react";
import "@mantine/core/styles.layer.css";
import "../App.scss";
import {
  LeftDynamicItemSetPairProvider,
  RightDynamicItemSetPairProvider,
} from "./DynamicItemSetPair";
import { TimerProvider } from "./Timer";

export enum Theme {
  Default = "default",
  Dark = "dark",
}

const themeStorageKey = "theme";

export function getTheme(): Theme {
  return (localStorage.getItem(themeStorageKey) || Theme.Default) as Theme;
}

const themes = {
  default: createTheme({
    primaryColor: "red",
  }),
  dark: createTheme({
    primaryColor: "red",
  }),
};

let themeChangeRef: MutableRefObject<(theme: Theme) => void>;

function setThemeClass() {
  const themeNode = document.getElementById("theme");
  if (!themeNode || !themeNode.className) {
    console.error("developers should have created an theme ID in the project");
    return;
  }
  themeNode.className = `${getTheme()}Theme`;
}

export function changeTheme(theme: Theme) {
  localStorage.setItem(themeStorageKey, theme.toString());
  themeChangeRef?.current && themeChangeRef.current(theme);
}

export const ShellProvider = ({ children }: PropsWithChildren) => {
  const [theme, setTheme] = useState<Theme>(getTheme());
  themeChangeRef = useRef(setTheme);
  setTimeout(() => {
    setThemeClass();
  }, 0);
  return (
    <div id="theme" className="defaultTheme">
      <LeftDynamicItemSetPairProvider>
        <RightDynamicItemSetPairProvider>
          <TimerProvider>
            <MantineProvider theme={themes[theme]}>{children}</MantineProvider>
          </TimerProvider>
        </RightDynamicItemSetPairProvider>
      </LeftDynamicItemSetPairProvider>
    </div>
  );
};
