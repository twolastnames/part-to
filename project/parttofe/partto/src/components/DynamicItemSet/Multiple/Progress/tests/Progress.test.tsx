import React from "react";
import { expect, test } from "@jest/globals";
import { render, screen } from "@testing-library/react";
import { ShellProvider } from "../../../../../providers/ShellProvider";
import { Progress } from "../Progress";

test("snapshot", () => {
  render(
    <ShellProvider>
      <Progress on={5} total={7} label={"11"} />
    </ShellProvider>,
  );
  const component = screen.getByTestId("LocalProgress");
  expect(component).toMatchSnapshot();
});
