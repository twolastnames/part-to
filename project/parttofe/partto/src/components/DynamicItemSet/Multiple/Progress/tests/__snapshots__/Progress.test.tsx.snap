// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<span
  data-testid="LocalProgress"
>
  <div
    class="m_b32e4812 mantine-RingProgress-root"
    data-testid="Progress"
    style="--rp-size: calc(3.5rem * var(--mantine-scale)); --rp-label-offset: calc(1rem * var(--mantine-scale));"
    title="Hello"
  >
    <svg
      class="m_d43b5134 mantine-RingProgress-svg"
    >
      <circle
        class="m_b1ca1fbf mantine-RingProgress-curve"
        cx="28"
        cy="28"
        fill="none"
        r="17.2"
        stroke-dasharray="15.438683897641276, 92.63210338584761"
        stroke-dashoffset="0"
        stroke-linecap="butt"
        stroke-width="8"
        transform="scale(1, -1) translate(0, -56)"
      />
      <circle
        class="m_b1ca1fbf mantine-RingProgress-curve"
        cx="28"
        cy="28"
        fill="none"
        r="17.2"
        stroke-dasharray="92.63210338584761, 15.438683897641276"
        stroke-dashoffset="108.07078728348888"
        stroke-linecap="butt"
        stroke-width="8"
        style="--curve-color: var(--secondary-color);"
      />
    </svg>
    <div
      class="m_b23f9dc4 mantine-RingProgress-label"
    >
      <div
        class="label"
      >
        11
      </div>
    </div>
  </div>
</span>
`;
